src/core/shape_inference/include/detection_output_shape_inference.hpp

```diff
@@ -0,0 +1,343 @@
+                              num_prior_boxes,
+                              "). Current value is: ",
+                              class_preds_pshape[1].get_length(),
+                              ".");
+        num_classes = (class_preds_pshape[1].get_length()) / num_prior_boxes;
+    }
+    if (num_classes == 0 && num_prior_boxes > 0 && box_logits_pshape.rank().is_static() &&
+        box_logits_pshape[1].is_static() && !attrs.share_location) {
+        NODE_VALIDATION_CHECK(op,
+                              (box_logits_pshape[1].get_length()) % (num_prior_boxes * 4) == 0,
+                              "Box logits second dimension must be a multiply of num_prior_boxes * 4 (",
+                              num_prior_boxes * 4,
+                              "). Current value is: ",
+                              box_logits_pshape[1].get_length(),
+                              ".");
+        num_classes = (box_logits_pshape[1].get_length()) / (num_prior_boxes * 4);
+    }
+    if (num_classes == 0 && num_prior_boxes > 0 && have_five_inputs && ad_box_preds_shape.rank().is_static() &&
+        ad_box_preds_shape[1].is_static() && !attrs.share_location) {
+        NODE_VALIDATION_CHECK(op,
+                              (ad_box_preds_shape[1].get_length()) % (num_prior_boxes * 4) == 0,
+                              "Additional box predictions second dimension must be a multiply of num_prior_boxes * 4 (",
+                              num_prior_boxes * 4,
+                              "). Current value is: ",
+                              ad_box_preds_shape[1].get_length(),
+                              ".");
+        num_classes = ad_box_preds_shape[1].get_length() / (num_prior_boxes * 4);
+    }
+}
+
+template <class T>
+void shape_infer_base(const DetectionOutputBase* op,
+                      const DetectionOutputBase::AttributesBase& attrs,
+                      const std::vector<T>& input_shapes,
+                      std::vector<T>& output_shapes,
+                      int64_t attribute_num_classes) {
+    using dim_t = typename std::iterator_traits<typename T::iterator>::value_type;
+    using val_type = typename dim_t::value_type;
+
+    NODE_VALIDATION_CHECK(op, (input_shapes.size() == 3 || input_shapes.size() == 5) && output_shapes.size() == 1);
+
+    auto& ret_output_shape = output_shapes[0];
+    ret_output_shape.resize(4);
+
+    const auto& box_logits_pshape = input_shapes[0];
+    const auto& class_preds_pshape = input_shapes[1];
+    const auto& proposals_pshape = input_shapes[2];
+
+    val_type num_classes = 0;
+    val_type num_prior_boxes = 0;
+    dim_t dim_num_images{};
+    bool dim_num_images_updated = false;
+
+    if (dynamic_cast<const ov::op::v8::DetectionOutput*>(op)) {
+        ov::op::util::compute_num_classes(op, attrs, input_shapes, num_classes, num_prior_boxes);
+    } else {
+        num_classes = static_cast<val_type>(attribute_num_classes);
+    }
```

Comment:
please get rid of this dynamic cast. It seems unnecessary to me. split shape infer into two if it's needed.